juju/juju
# feat: pool atomic context

The atomic context can be pooled to prevent the creation of objects on the heap. There was originally a TODO about doing this, but until the state base was finalized it was left to do until later. There are a few services already using the atomic setup, so it feels like we should do this before it's forgotten.

There should be no changes to the testing, this is just an implementation detail and should be transparent to the tests.

This was observed when testing https://github.com/juju/juju/pull/18007





## Checklist



- [x ] Go unit tests, with comments saying what you're testing

## QA steps

Existing tests should pass.

Testing adding units to ensure no regressions have been performed is a good test also.

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy ubuntu
$ juju remove-application ubuntu
```

## Links


**Jira card:** https://warthogs.atlassian.net/browse/JUJU-6625
